processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java

```diff
@@ -263,8 +299,17 @@ public ValueType defaultType()
         final IndexedInts row = selector.getRow();
 
         if (row.size() == 1) {
-          final String key = selector.lookupName(row.get(0));
-          return index.find(key).iterator();
+          int dimensionId = row.get(0);
+          final IntList rowNumbers;
+          if (selector.getValueCardinality() == DimensionDictionarySelector.CARDINALITY_UNKNOWN) {
```

Comment:
This won't change from row to row, so it's likely going to be better to check it outside hot code (i.e. return a different `Supplier<IntIterator>` based on the result of calling `selector.getValueCardinality()`).